internal/controllers/core/kubernetesdiscovery/portforwards.go

```diff
@@ -180,3 +189,32 @@ func isBetterPortForwardPod(podA, podB *v1alpha1.Pod) bool {
 	// Use the name as a tie-breaker.
 	return podA.Name > podB.Name
 }
+
+func warnDeprecatedImplicitForwards(ctx context.Context, kd *v1alpha1.KubernetesDiscovery, pf *v1alpha1.PortForward) {
+	if kd == nil || pf == nil {
+		return
+	}
+
+	var implicit []string
+	for _, pft := range kd.Spec.PortForwardTemplateSpec.Forwards {
+		if pft.ContainerPort != 0 {
+			continue
+		}
+
+		for _, f := range pf.Spec.Forwards {
+			if pft.LocalPort == f.LocalPort && f.LocalPort != f.ContainerPort {
+				implicit = append(implicit,
+					fmt.Sprintf("  – %d ⇢ %d:%d", f.LocalPort, f.LocalPort, f.ContainerPort))
+			}
+		}
+	}
+
+	if len(implicit) != 0 {
+		logger.Get(ctx).Warnf(
+			"One or more port forwards did not specify a container port, and no exposed container port matched the local port.\n"+
```

Comment:
a couple independent questions:

- maybe we should _always_ print the port forward mapping in the logs? i think that would have helped with the original complaint.

- I'm not sure most users are going to understand what to do with this error message...i would be much more explicit about what needs to change. Something like:

`k8s_resource(name=x, port_forward='8000')` currently maps `localhost:8000` to port 8001 in your container.

Tilt vX.Y.Z will change this default! k8s_resource(port_forward='8000')` will map `localhost:8000` to port 8000 in your container.

To keep your project working, change your Tiltfile to `k8s_resource(name=x, port_forward='8000:8001')`